### ISSUE TITLE
Center the square glyphs

### ISSUE THREAD

**Issue by boltlessengineer** [wez/wezterm#2818]
# Problem describe

This issue covers a wider range of problems related to #2794.

Wezterm just prints nerd font icons on bottom-left. This makes square glyphs lean to left side and resized glyphs lean to bottom.

### When allowing width overflow

![image](https://user-images.githubusercontent.com/60088301/204997702-668184e0-0e5e-4c06-8b8e-9289089cd2d6.png)

When `glyph width` is smaller than `2*cell_width`, the glyph looks like it is lean to left side. It would be great if wezterm automatically gives x_offset to center the square glyph in 2-cell block.

### When not allowing width overflow (previously described in #2794)

![image](https://user-images.githubusercontent.com/60088301/204997731-b728b872-d6a5-4352-834e-d02b698ebc5a.png)

When not allowing width overflow, wezterm just scale the font down and align it to bottom. It would be better if wezterm vertically centers scaled fonts.

**Comment by wez** [comment]
> Wezterm just prints nerd font icons on bottom-left.

No, that assumption is incorrect. Wezterm uses harfbuzz to correctly shape and compute placement information based on the metrics contained in the font program.

When working with font fallback, wezterm aligns the baselines of the fonts based on their metrics.

In general, wezterm doesn't know anything about the composition of a glyph; it only knows where to position the baseline vs. the bottom of the cell box and then asks harfbuzz to position it, and freetype to rasterize it.  It may surprise you, but often the computed positioning metrics are for rectangles that are much smaller than the cell boundaries and often have x and y offsets that are not aligned to the cell corners or baseline.

Font metrics are very abstract and the interpretation of them varies from font family to font family; it is up to the font designer to decide how their glyphs fill up the [EM-Square](http://designwithfontforge.com/en-US/The_EM_Square.html).  That makes it especially difficult to reason about text alignment and placement between different fonts.  You can read a bit more about this here: https://wezfurlong.org/wezterm/config/lua/wezterm/font_with_fallback.html#dealing-with-different-fallback-font-heights

I understand the nature of your request, but it is very difficult for an application to make the right kind of decision about overriding the information provided by the font program itself.

I'm willing to consider adding some options similar in nature to the `scale` parameter of `wezterm.font_with_fallback` that would allow the user to employ their own judgement on fine tuning the positioning for their particular set of fonts.

Also, FWIW: the nerd fonts have a history of shipping with somewhat mangled font metrics, so I do not want to hard code logic in wezterm based on observations made with bad font metrics.

**Comment by boltlessengineer** [comment]
Thank you for your detail explanation.

> wezterm doesn't know anything about the composition of a glyph

What do you mean _composition_? Doesn't wezterm knows about width and height of a glyph? I thought that was how `allow_square_glyphs_to_overflow_width` option works.

> I do not want to hard code logic in wezterm based on observations made with bad font metrics

I don't know about Rust that much, but I don't think this would be a hard coding. Wezterm already seems to detect square glyphs and behaves differently for those glyphs (like `allow_square_glyphs_to_overflow_width` option). Can't we just add a little logic for modifying x and y offsets to center the glyphs if those are _square glyphs_?

I'm sorry if I misunderstood. My English skill is quite bad.

**Comment by dpc** [comment]
Was bothering me a lot, so I sloped it: eb8f675799977fdb412c14a9d32210df60d1b6af
